rfcs/text/0003_handler_interface.md

```diff
@@ -116,12 +124,75 @@ http.router.route({
 });
 ```
 
+## Types
+
+While services that implement this pattern will not be able to define a static
+type, plugins should be able to reopen a type to extend it with whatever context
+it provides. This allows the `registerContext` function to be type-safe.
+For example, if the HTTP service defined a setup type like this:
+
+```ts
+// http_service.ts
+interface RequestContext {
+  [contextName: string]: unknown;
+}
+
+interface HttpSetup {
+  // ...
+
+  registerRequestContext<T extends keyof RequestContext>(
+    contextName: T,
+    provider: (request: Request) => RequestContext[T]
+  ): void;
+
+  // ...
+}
+```
+
+A consuming plugin could extend the `RequestContext` to be type-safe like this:
+
+```ts
+// my_plugin/server/index.ts
+import { RequestContext } from '../../core/server';
+
+// The plugin *has to* add a new property to the RequestContext interface from
+// core to represent whatever type its context provider returns. This will be
+// available to any module that imports this type and will ensure that the
+// registered context provider returns the expected type.
+declare module "../../core/server" {
+  interface RequestContext {
+    myPlugin?: {  // should be optional because this plugin may be disabled.
+      getFoo(): string;
+    }
+  }
+}
+
+class MyPlugin {
+  setup(core) {
+    // This will be type-safe!
+    core.http.registerRequestContext('myPlugin', (request) => ({
+      getFoo() { return 'foo' }
+    }))
+  }
+};
+```
+
 # Drawbacks
 
-- Since the service names are passed as strings but accessed as properties,
-  it's unclear how best to type these context objects. We can expose types at
-  the core and plugin level, but consumers of those types might need to compose
-  them together manually.
+- Since the context properties that are present changes if plugins are disabled,
+  they are all marked as optional properties which makes consuming the context
+  type awkward. We can expose types at the core and plugin level, but consumers
+  of those types might need to define which properties are present manually to
+  match their required plugin dependencies. Example:
+  ```ts
+  type RequiredDependencies = 'elasticsearch';
+  type OptionalDependencies = 'data' | 'telemetry';
+  type KnownContext = Required<Pick<RequestContext, RequiredDependencies>> & Pick<RequestContext, OptionalDependencies>;
+  // => { elasticsearch: ESClient, data?: Data, telemetry?: Telemetry };
+  ```
+- Extending types with `declare module` merging is not a typical pattern for
```

Comment:
this pattern is already widely used in Kibana
https://github.com/restrry/kibana/blob/31bbb4c8e1d35247b73a575de26e08b2ea5f4542/src/legacy/server/kbn_server.d.ts#L53
https://github.com/restrry/kibana/blob/31bbb4c8e1d35247b73a575de26e08b2ea5f4542/x-pack/plugins/infra/types/eui.d.ts#L22